File: EPRO.Library.v3.5/EPRO.Library.v3.5/Modules/basObjectCloner.vb
Imports System.Runtime.Serialization
Imports System.Runtime.CompilerServices
Imports System.Runtime.Serialization.Formatters.Binary
Imports System.IO

Namespace Modules
    Public Module basObjectCloner

        <Extension()> _
        Public Function Clone(Of T)(source As T) As T

            If Not GetType(T).IsSerializable() Then Throw New ArgumentException("The type must be serializable.", "source")

            If (Object.ReferenceEquals(source, Nothing)) Then Return Nothing


            Dim formatter As IFormatter = New BinaryFormatter()
            Dim stream As Stream = New MemoryStream()
            Using (stream)

                formatter.Serialize(stream, source)
                stream.Seek(0, SeekOrigin.Begin)
                Return CType(formatter.Deserialize(stream), T)
            End Using

        End Function

    End Module

End Namespace

File: EPRO.Library.v3.5/EPRO.Library.v3.5/Modules/basObjectCloner.cs
using System;
using System.IO;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;

namespace ELibrary.Standard.Modules
{
    public static class basObjectCloner
    {
        public static T Clone<T>(this T source)
        {
            if (!typeof(T).IsSerializable)
                throw new ArgumentException("The type must be serializable.", "source");
            if (ReferenceEquals(source, null))
                return default;
            IFormatter formatter = new BinaryFormatter();
            Stream stream = new MemoryStream();
            using (stream)
            {
                formatter.Serialize(stream, source);
                stream.Seek(0L, SeekOrigin.Begin);
                return (T)formatter.Deserialize(stream);
            }
        }
    }
}